src/plugins/intel_gpu/include/intel_gpu/runtime/utils.hpp

```diff
@@ -183,4 +189,97 @@ inline std::string to_string(const T& v) {
 /// @}
 /// @endcond
 /// @}
+
+/// @brief LRU cache which remove the least recently used data when cache is full.
+template<typename TypeK, typename TypeD>
+class LRUCache {
+public:
+    struct CacheEntry {
+        TypeD   data;
+        size_t  size;   // the size of data. it is necessary to check total data size is limited to capacity
+    };
+
+private:
+    std::list<TypeK> order;
+    std::unordered_map<TypeK, std::pair<typename std::list<TypeK>::iterator, LRUCache::CacheEntry>> cache_entry_map;
+    size_t curr_data_size;
+    const size_t capacity;
+
+    std::mutex mtx;
+
+public:
+    explicit LRUCache(size_t caps) : curr_data_size(0), capacity(caps) {}
+
+    ~LRUCache() {
+        clear();
+    }
+
+    TypeD get_lru_element() {
+        if (order.size()) {
+            return cache_entry_map[order.back()].second.data;
+        } else {
+            return nullptr;
+        }
+    }
+    std::tuple<TypeD, bool> get(TypeK key, std::function<LRUCache::CacheEntry(void)> create_new_data = nullptr, bool* last_element_popped = nullptr) {
```

Comment:
Actually this was mimicing processor cache where the entry is registered when it is accessed. But I have no strong objection to split it to multiple methods. Maybe it requires more responsibility to the caller side, but it will have more flexibliity

Reply:
Modified it. could you please review it again?